src/meta/backup_engine.cpp

```diff
@@ -215,37 +212,53 @@ void backup_engine::on_backup_reply(error_code err,
     // if backup failed, receive ERR_LOCAL_APP_FAILURE;
     // backup not completed in other cases.
     // see replica::on_cold_backup() for details.
-    int32_t partition = pid.get_partition_index();
-    if (err == dsn::ERR_OK && response.err == dsn::ERR_OK &&
-        response.progress == cold_backup_constant::PROGRESS_FINISHED) {
+    if (err != ERR_OK || response.err != ERR_OK) {
+        if (response.err == ERR_LOCAL_APP_FAILURE) {
+            dcheck_eq(response.pid, pid);
+            dcheck_eq(response.backup_id, _cur_backup.backup_id);
+
+            derror_f("backup_id({}): backup for partition {} failed, error message: {}, "
+                     "response.err: {}",
+                     _cur_backup.backup_id,
+                     pid.to_string(),
+                     err.to_string(),
+                     response.err.to_string());
+            zauto_lock l(_lock);
+            // if one partition fail, the whole backup plan fail.
+            _is_backup_failed = true;
+            _backup_status[pid.get_partition_index()] = backup_status::FAILED;
+            return;
+        } else {
+            // backup not completed, retry sending requests
+            dwarn_f("backup_id({}): send backup request to server {} failed, rpc error: {}, "
```

Comment:
These code is same with L261~L275, could you refactor it?

Reply:
It's not the same

Reply:
I found there are a lot of same code between them

Reply:
the logic is different